File: ruby/life.rb
require "set"

showWork = false
generations = 1000

Change = Struct.new(:how, :pos)

def applyUpdates(liveSet, updates)
  toLive = updates.filter_map{|c| c.pos if c.how == 'live'}
  toDie = updates.filter_map{|c| c.pos if c.how == 'die'}
  (liveSet | toLive) - toDie
end

def boundingBox(liveSet)
  xs = liveSet.map{|p| p[0]}
  ys = liveSet.map{|p| p[1]}
  [[xs.min, ys.min], [xs.max, ys.max]]
end

def printBoard(liveSet)
  bbox = boundingBox(liveSet)
  min = bbox[0]
  max = bbox[1]
  # clear the screaen
  print "\e[2J"
  # move to top-left corner
  print "\e[;H"
  print ""
  for y in max[1].downto(min[1]) do
    for x in (min[0] .. max[0]) do
      if liveSet.member? [x,y]
        print "@"
      else
        print " "
      end
    end
    puts ""
  end
  sleep 1/30.0
end

def eight(pos)
  (-1..1).flat_map{
    |x| (-1..1).filter_map{
      |y| [pos[0] + x, pos[1] + y] if x != 0 or y != 0
    }
  }
end

def computeAffected(updates)
  Set.new(updates.flat_map{|change| eight(change.pos)})
end

def neighborCount(liveSet, pos)
  eight(pos).filter_map{|n| 1 if liveSet.member? n}.sum
end

def computeChange(liveSet, pos, count)
  if count == 2
    return nil
  end
  alive = liveSet.member? pos
  if count == 3
    if !alive
      return Change.new('live', pos)
    end
  else
      if alive
        return Change.new('die', pos)
      end
  end
  nil
end

def computeChanges(liveSet, affected)
  affected.map{
    |pos|
    computeChange(liveSet, pos, neighborCount(liveSet, pos))
  }.select{|c| not c.nil?}
end

def generation(showWork, n, liveSet, updates)
  if n > 0
    newSet = applyUpdates(liveSet, updates)
    if showWork
      printBoard newSet
    end
    affected = computeAffected(updates)
    newChanges = computeChanges(newSet, affected)
    generation(showWork, n - 1, newSet, newChanges)
  end
end


def run(initialChanges, showWork, generations)
  start = Time.now
  generation(showWork, generations, Set.new, initialChanges)
  diff = Time.now - start
  puts "%.2f generations / sec" % (1000 / diff)
end

r_pentomino = [
  [0, 0], [0, 1], [1, 1], [-1, 0], [0, -1]
].map{ |pos| Change.new('live', pos)}
  
if showWork
  run(r_pentomino, showWork, generations)
else
  for i in 1..5
    run(r_pentomino, showWork, generations)
  end
end

File: elixir/life.exs
defmodule Life do

  def applyUpdates(liveSet, updates) do
      live = MapSet.new(Keyword.get_values(updates, :live))
      die = MapSet.new(Keyword.get_values(updates, :die))
      MapSet.difference(MapSet.union(liveSet, live), die)
  end

  def eight(pos) do
    {posx, posy} = pos
    nine = for x <- -1..1, y <- -1..1, do: if x != 0 or y != 0, do: {posx + x, posy + y}
    Enum.filter(nine, &!is_nil(&1))
  end

  def neighbors(changes) do
    MapSet.new(Enum.flat_map(Keyword.values(changes), &eight(&1)))
  end

  def neighborCount(liveSet, pos) do
    Enum.count(Enum.filter(eight(pos), &alive?(liveSet, &1)))
  end

  def rules(liveSet, pos) do
    count = neighborCount(liveSet, pos)
    case count do
      2 -> nil
      3 -> if !alive?(liveSet, pos), do: {:live, pos}
      _ -> if alive?(liveSet, pos), do: {:die, pos}
    end
  end

  def keep(collection, function) do
    Enum.filter(Enum.map(collection, function), fn (n) -> !is_nil(n) end)
  end

  # process collection, in chunks, in parallel
  # removes nil responses, via keep, above
  def pmap(chunkSize, collection, function) do
    tasks = for chunk <- Enum.chunk_every(collection, chunkSize),
      do: Task.async(fn -> keep(chunk, function) end),
      into: []
    Enum.concat(for t <- tasks, do: Task.await(t))
  end

  def computeChanges(parallelism, liveSet, affected) do
    chunk = max(div(MapSet.size(affected), parallelism), 1)
    pmap(chunk, affected, &rules(liveSet, &1))
  end

  def boundingBox(liveSet) do
    {minx, maxx} = Enum.min_max(Enum.map(liveSet, fn ({x,_}) -> x end))
    {miny, maxy} = Enum.min_max(Enum.map(liveSet, fn ({_,y}) -> y end))
    {{minx, miny}, {maxx, maxy}}
  end

  def alive?(liveSet, pos) do
    MapSet.member?(liveSet, pos)
  end

  def printRow(liveSet, y, minx, maxx) do
    IO.puts(
      for x <- minx..maxx, do: if alive?(liveSet, {x,y}), do: '@', else: ' '
    )
  end

  def clearScreen() do
    IO.write("\e[2J")
    IO.write("\e[;H")
  end

  def printBoard(liveSet) do
    clearScreen()
    {{minx, miny}, {maxx, maxy}} = boundingBox(liveSet)
    for y <- maxy..miny, do: printRow(liveSet, y, minx, maxx)
    :timer.sleep(div(1000, 30))
  end
  
  def generation(n, showWork, liveSet, updates, parallelism) do
    if n > 0 do
      newSet = applyUpdates(liveSet, updates)
      if showWork do
	printBoard(newSet)
      end
      affected = neighbors(updates)
      newChanges = computeChanges(parallelism, newSet, affected)
      generation(n - 1, showWork, newSet, newChanges, parallelism)
    end
  end

  def life(n, showWork, parallelism) do
    start = :os.system_time(:millisecond)
    r_pentomino = Enum.map(
      [{0, 0}, {0, 1}, {1, 1}, {-1, 0}, {0, -1}],
      & {:live, &1}
    )
    generation(n, showWork, MapSet.new(), r_pentomino, parallelism)
    diff = :os.system_time(:millisecond) - start
    IO.puts(to_string(n / (diff / 1000.0)) <> " generations per sec")
  end
end

if false do
  Life.life(300, true, 1)
else
  for _ <- 1..5 do
      Life.life(1000, false, 8)
  end
end
